kube-runtime/src/coordination.rs

```diff
@@ -0,0 +1,603 @@
+        // Attempt to fetch the target lease, updating our last observed info on the lease.
+        let lease_opt = timeout(self.config.api_timeout, self.api.get_opt(&self.config.name))
+            .await
+            .map_err(|_err| Error::TimeoutError)?
+            .map_err(Error::ClientError)?;
+
+        // Lease info returned from the API, update observation info.
+        if let Some(lease) = lease_opt {
+            self.update_state(Some(lease));
+            return Ok(());
+        }
+        Ok(())
+    }
+
+    /// Attempt to acquire or renew the target lease.
+    #[tracing::instrument(level = "debug", skip_all, err)]
+    #[allow(clippy::cast_possible_truncation)]
+    async fn try_acquire_or_renew(&mut self) -> Result<()> {
+        tracing::debug!("try_acquire_or_renew");
+        // 1. Fetch the current state of the lease if following or standby.
+        if matches!(&self.state, State::Following { .. } | State::Standby) {
+            self.k8s_fetch_lease_and_update().await?;
+        }
+
+        // 2. Determine what type of update needs to be made to the lease.
+        // If following, and the lease is not expired (according to our own local time records), then no-op.
+        if matches!(&self.state, State::Following { .. }) && !self.is_lease_expired() {
+            return Ok(());
+        }
+        // Else, we need to patch the lease. Build up changeset.
+        let now = chrono::Utc::now();
+        let mut lease = self.state.get_lease().cloned().unwrap_or_default();
+        lease
+            .meta_mut()
+            .name
+            .get_or_insert_with(|| self.config.name.clone());
+        lease
+            .meta_mut()
+            .namespace
+            .get_or_insert_with(|| self.config.namespace.clone());
+        let spec = lease.spec.get_or_insert_with(Default::default);
+        spec.lease_duration_seconds = Some(self.config.lease_duration.as_secs_f32() as i32);
+        spec.renew_time = Some(MicroTime(chrono::Utc::now()));
+        if matches!(&self.state, State::Following { .. } | State::Standby) {
+            spec.holder_identity = Some(self.config.identity.clone());
+            spec.acquire_time = Some(MicroTime(now));
+            spec.lease_transitions = Some(spec.lease_transitions.map_or(0, |val| val + 1));
+        }
+        lease.metadata.managed_fields = None; // Can not pass this along for update.
+
+        // 3. Now we need to create or patch the lease in K8s with the updated lease value here.
+        let lease_res = if let State::Standby = &self.state {
+            timeout(
+                self.config.api_timeout,
+                self.api.create(&Default::default(), &lease),
+            )
+            .await
+        } else {
+            let mut params = PatchParams::apply(&self.config.manager);
+            params.force = true; // This will still be blocked by the server if we do not have the most up-to-date lease info.
```

Comment:
I would note that this the recommended behaviour for controllers; https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts

Reply:
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#concurrency-control-and-consistency only mentions that `resourceVersion` is validated for `PUT` (`.replace()`) requests, and https://kubernetes.io/docs/reference/using-api/server-side-apply/ doesn't mention it at all.

I don't believe there is any official documentation (that I can find, anyway..) that hints towards SSA being a safe way to lock things.

Reply:
> I would note that this [is] the recommended behaviour for controllers

Will do.

> [...]/server-side-apply/ doesn't mention it at all.

You are correct that the SSA docs don't mention this; however, the API conventions docs do, specifically under the CC&C section:

> When a record is about to be updated, it's version is checked against a pre-saved value, and if it doesn't match, the update fails with a StatusConflict (HTTP status code 409).

The main difference with SSA is that a resourceVersion is not required, and updates will still be accepted if no resourceVersions is specified. _We are not leveraging that rule here._

This all falls under the topic of idempotency and CC&C. Here, our apply will still be rejected if the resourceVersions do not match (we still pass that info along). And the create immediately above it (the first branch of the if block) will be rejected if the object already exists. So we are g2g here.

Reply:
> The main difference with SSA is that a resourceVersion is not required, and updates will still be accepted if no resourceVersions is specified. We are not leveraging that rule here.

No, the docs only mention respecting `resourceVersion` at all for PUT requests:

> If resourceVersion is included with the PUT operation the system will verify that there have not been other successful mutations to the resource during a read/modify/write cycle, by verifying that the current value of resourceVersion matches the specified value.

I can't seem to find the reference now, but I remember an older version of the documentation explicitly mentioning that `resourceVersion` was _never_ validated for PATCH requests (which SSA is one kind of).

Reply:
> No, the docs only mention respecting resourceVersion at all for PUT requests:

Not quite true. I think we can all agree that the docs could be improved in terms of disambiguating issues like this; however, the paragraph above the line you quoted states:

> When a record is about to be updated, it's version is checked against a pre-saved value, and if it doesn't match, the update fails with a StatusConflict (HTTP status code 409).

That is a broad statement. If the resourceVersion is present in the request, and the object is about to be updated, then the versions are checked.

To back this up with some proof, see the following code.

```bash
# Create a cluster, sets the default context used below (here k8s v1.25).
kind create cluster --name test
```

```rust
//! Concurrency Control & Consistency

use anyhow::{Context, Result};
use k8s_openapi::api::coordination::v1::Lease;

use kube::{
    api::{Api, Patch, PatchParams, PostParams},
    Resource,
};

#[tokio::main]
async fn main() -> Result<()> {
    println!("connecting to cluster");
    let client = kube::Client::try_default().await.context("error initializing K8s client")?;
    let api: Api<Lease> = Api::namespaced(client, "default");

    // Build a new lease.
    println!("building new lease");
    let mut new_lease = Lease::default();
    let meta = new_lease.meta_mut();
    meta.name = Some("test".into());
    meta.namespace = Some("default".into());

    // Creat the new lease.
    println!("creating new lease in k8s");
    let postp = PostParams::default();
    let mut lease = api.create(&postp, &new_lease).await.context("error creating new lease")?;
    let ver = lease.meta().resource_version.clone();

    // Patch the lease.
    println!("patching the lease");
    let spec = lease.spec.get_or_insert_with(Default::default);
    spec.holder_identity = Some("me".into());
    lease.meta_mut().resource_version = Some("1".into());
    anyhow::ensure!(lease.meta().resource_version != ver);
    let mut patchp = PatchParams::apply("test-manager");
    patchp.force = true;
    api.patch("test", &patchp, &Patch::Apply(lease)).await.context("error applying patch")?;

    Ok(())
}
```

If you execute the above code, everything works until that last `api.patch`, and the error it returns:

> ApiError: Operation cannot be fulfilled on leases.coordination.k8s.io "test": the object has been modified; please apply your changes to the latest version and try again: Conflict (ErrorResponse { status: "Failure", message: "Operation cannot be fulfilled on leases.coordination.k8s.io \"test\": the object has been modified; please apply your changes to the latest version and try again", reason: "Conflict", code: 409 })

This is inline with the first paragraph of the CC&C section of the API conventions doc. This is the expectation which the code in this PR uses.

Hopefully we can move on. Please let me know if I have overlooked anything.

Reply:
Ok, I've addressed the initial request here about documented the recommended behavior for controllers. I'll keep this thread open in case there are other items to discuss related to CC&C.

Reply:
Upon re-reading my response here, I see that there may be one thing which is not entirely clear. In the code block I shared, specifically these lines:
```rust
lease.meta_mut().resource_version = Some("1".into());
anyhow::ensure!(lease.meta().resource_version != ver);
```
this code is changing the resource_version to be some different value so that it does not match what K8s currently has stored. The second line asserts that the version k8s has is indeed different from the version we've changed it to.

Then, when the code is executed, we see that the patch fails as expected according to the API conventions doc. Just wanted to clarify.

Reply:
Any additional thoughts here folks? Should I mark as resolved?